### PR TITLE
LM training, support for SentencePiece and HF tokenizers 

### PR DESCRIPTION
**Pull request by AlexGrinch**
1. Added support for SentencePiece and HF tokenizers when training LMs to fix the bug (https://github.com/NVIDIA/NeMo/issues/2544#issuecomment-899922327)
2. Removed obsolete text post-processing in AutoTokenizer

### CODE REVIEW

File: examples/nlp/machine_translation/create_tarred_monolingual_dataset.py [hunk 1]
```diff
@@ -19,8 +19,10 @@
 
 if __name__ == '__main__':
     parser = argparse.ArgumentParser(description='NMT dataset pre-processing')
-    parser.add_argument('--tokenizer_model', type=str, required=True, help='Path to tokenizer model')
-    parser.add_argument('--tokenizer_name', type=str, default='yttm', help='BPE Tokenizer Name, Options: [yttm]')
+    parser.add_argument(
+        '--tokenizer_name', type=str, default='yttm', help='Supports yttm, sentencepiece and HuggingFace tokenizers',
+    )
+    parser.add_argument('--tokenizer_model', type=str, default=None, help='Path to tokenizer model')
```
Comment (MaximumEntropy) [inline examples/nlp/machine_translation/create_tarred_monolingual_dataset.py:25]: I think having the tokenizer_model is necessary and it isn't possible to train a tokenizer on the provided data right now correct? If so, I think we should make this a required argparse field.

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py [hunk 2]
```diff
@@ -203,7 +172,7 @@ def text_to_tokens(self, text):
 
     def tokens_to_text(self, tokens):
         text = self.tokenizer.convert_tokens_to_string(tokens)
-        return remove_spaces(handle_quotes(text.strip()))
+        return text
```
Comment (MaximumEntropy) [inline nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:175]: High-level question about changes to this entire file - are we sure it doesn't break things for others using HF tokenizers?

**Review by MaximumEntropy — APPROVED** [review]
LGTM. Just a couple of minor comments/questions.

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py (hunk 2 above)
Comment (AlexGrinch) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:175]: The initial reason for all this post-processing was to correctly process all the punctuation after bert-base-uncased detokenization 2 years ago when we tried to fine-tune it on En-De NMT. After that the tokenization/detokenization in HF improved significantly.

File: examples/nlp/machine_translation/create_tarred_monolingual_dataset.py (hunk 1 above)
Comment (AlexGrinch) [reply examples/nlp/machine_translation/create_tarred_monolingual_dataset.py:25]: The reason for removing `required` flag is to support HF tokenisers out of the box. They are usually created by passing corresponding HF model name to `tokenizer_name`.

Alternative design choice is to pass something like `hf` or `hugging_face` as `tokenizer_name` and expecting `tokenizer_model` to be the name of the corresponding HF model.

File: nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py (hunk 2 above)
Comment (MaximumEntropy) [reply nemo/collections/common/tokenizers/huggingface/auto_tokenizer.py:175]: I see, makes sense. Thanks for clarifying!

**Pull request merged**